### ISSUE TITLE
Improper error type is returned when a runtime API method is not available

### ISSUE THREAD

**Issue by s0me0ne-unkn0wn** [paritytech/substrate#13344]
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Description of bug

When a non-existing runtime API method call is requested, [the error returned](https://github.com/paritytech/substrate/blob/6f72780b87e41a025722284a52a7d85f41958a82/client/executor/wasmtime/src/instance_wrapper.rs#L202) is of type [`Other(String)`](https://github.com/paritytech/substrate/blob/6f72780b87e41a025722284a52a7d85f41958a82/client/executor/common/src/error.rs#L113).

At that point, the information about why the error happened is lost, and that doesn't leave the caller any other chance besides treating it as a [`RuntimeApiError::Execution`](https://github.com/paritytech/polkadot/blob/dfa8c5ae321e0523f88fdc43afc6d743812ee49e/node/core/runtime-api/src/lib.rs#L376). That is, the caller knows that an error has happened when trying to execute the runtime call. Still, it doesn't know if that's a problem with the runtime call logic itself or if the current runtime does not support the runtime call requested.

That may cause problems with the current code, as well as with the future one. For example, the newer disputes logic [may rely upon the fact](https://github.com/paritytech/polkadot/blob/dfa8c5ae321e0523f88fdc43afc6d743812ee49e/node/core/provisioner/src/disputes/prioritized_selection/mod.rs#L478) the executor should return `NotSupported` error instead of the `Execution` one. The `NotSupported` error variant is quite intuitive and incentives the caller to use it to check if the runtime call is supported at all; nevertheless, the only case when it's really used is checking [if the runtime supports runtime API versioning](https://github.com/paritytech/polkadot/blob/dfa8c5ae321e0523f88fdc43afc6d743812ee49e/node/core/runtime-api/src/lib.rs#L396).

The proposal is to return the [`MethodNotFound` variant](https://github.com/paritytech/substrate/blob/6f72780b87e41a025722284a52a7d85f41958a82/client/executor/common/src/error.rs#L36) and to further map it to the [`RuntimeApiError::NotSupported`](https://github.com/paritytech/polkadot/blob/dfa8c5ae321e0523f88fdc43afc6d743812ee49e/node/subsystem-types/src/errors.rs#L37) to make things more consistent.

**Comment by bkchr** [comment]
Yes the error variant should be improved. However, the caller already can use https://docs.rs/sp-api/latest/sp_api/trait.ApiExt.html#tymethod.api_version to check if the runtime supports the expected version.

**Comment by s0me0ne-unkn0wn** [comment]
Yes, that's totally legit. But I mean, why do two runtime calls instead of one? Also, it's pretty counter-intuitive to get the `Execution` error instead of the `NotSupported` one when the runtime call is not supported.

**Comment by s0me0ne-unkn0wn** [comment]
Okay, I get it. But here is some further investigation:

The network was running rococo-9360 runtime, which declares its support for parachain host API version 3. But 9360 is still missing the `session_executor_params` call, which will (hopefully) be introduced in 9380 without the API version bump.

So version checking is a valuable mechanism (and it is used and is working), but it is unreliable for staging changes unless they involve API version bump.

**Comment by ordian** [comment]
Once we add some method as stable (e.g. `disputes` as `api_version(3)`), there's a risk someone will cut a release with that version (this is what happened with Rococo/Versi). Which is why I suggested to add additional methods as version 4.

**Comment by s0me0ne-unkn0wn** [comment]
> Once we add some method as stable (e.g. `disputes` as `api_version(3)`)

But I believe `api_version(3)` is not stable yet, it's only implemented by Rococo and Westend.

We wanted to stabilize it because of `SessionInfo` changes but reverted that later.
